probe/resolver.go

```diff
@@ -15,93 +15,115 @@ var (
-	port     string
-}
+type target struct{ host, port string }
+
+func (t target) String() string { return net.JoinHostPort(t.host, t.port) }
 
-// NewResolver starts a new resolver that periodically
-// tries to resolve peers and the calls add() with all the
-// resolved IPs.  It explictiy supports hostnames which
-// resolve to multiple IPs; it will repeatedly call
-// add with the same IP, expecting the target to dedupe.
-func newStaticResolver(peers []string, add func(string)) staticResolver {
+// newStaticResolver periodically resolves the targets, and calls the set
+// function with all the resolved IPs. It explictiy supports targets which
+// resolve to multiple IPs.
+func newStaticResolver(targets []string, set func(target string, endpoints []string)) staticResolver {
 	r := staticResolver{
-		quit:  make(chan struct{}),
-		add:   add,
-		peers: prepareNames(peers),
+		targets: prepare(targets),
+		set:     set,
+		sema:    newSemaphore(maxConcurrentLookup),
+		quit:    make(chan struct{}),
 	}
 	go r.loop()
 	return r
 }
 
-func prepareNames(strs []string) []peer {
-	var results []peer
-	for _, s := range strs {
-		var (
-			hostname string
-			port     string
-		)
-
-		if strings.Contains(s, ":") {
-			var err error
-			hostname, port, err = net.SplitHostPort(s)
-			if err != nil {
-				log.Printf("invalid address %s: %v", s, err)
-				continue
-			}
-		} else {
-			hostname, port = s, strconv.Itoa(xfer.AppPort)
-		}
-
-		results = append(results, peer{hostname, port})
-	}
-	return results
-}
-
 func (r staticResolver) loop() {
-	r.resolveHosts()
+	r.resolve()
 	t := tick(time.Minute)
 	for {
 		select {
 		case <-t:
-			r.resolveHosts()
-
+			r.resolve()
 		case <-r.quit:
 			return
 		}
 	}
 }
 
-func (r staticResolver) resolveHosts() {
-	for _, peer := range r.peers {
-		var addrs []net.IP
-		if addr := net.ParseIP(peer.hostname); addr != nil {
-			addrs = []net.IP{addr}
-		} else {
+func (r staticResolver) Stop() {
+	close(r.quit)
+}
+
+func prepare(strs []string) []target {
+	var targets []target
+	for _, s := range strs {
+		var host, port string
+		if strings.Contains(s, ":") {
 			var err error
-			addrs, err = lookupIP(peer.hostname)
+			host, port, err = net.SplitHostPort(s)
 			if err != nil {
+				log.Printf("invalid address %s: %v", s, err)
 				continue
 			}
+		} else {
+			host, port = s, strconv.Itoa(xfer.AppPort)
 		}
+		targets = append(targets, target{host, port})
+	}
+	return targets
+}
 
-		for _, addr := range addrs {
-			// For now, ignore IPv6
-			if addr.To4() == nil {
-				continue
-			}
-			r.add(net.JoinHostPort(addr.String(), peer.port))
+func (r staticResolver) resolve() {
+	for t, endpoints := range resolveMany(r.sema, r.targets) {
+		r.set(t.String(), endpoints)
+	}
+}
+
+func resolveMany(s semaphore, targets []target) map[target][]string {
+	result := map[target][]string{}
+	for _, t := range targets {
+		c := make(chan []string)
+		go func(t target) { s.p(); defer s.v(); c <- resolveOne(t) }(t)
```

Comment:
why are you doing the dns lookups on a separate goroutine?  I think its pretty safe to do these on this routine; go will timeout requests after 5 seconds if they don't get a reply.